crate_universe/Cargo.toml

```diff
@@ -21,7 +21,7 @@ default = ["cargo"]
 [dependencies]
 anyhow = "1.0.66"
 cargo_metadata = "0.15.2"
-cargo_toml = "0.13.0"
+cargo_toml = "0.14.0"
```

Comment:
When updating dependencies here, I would recommend running `bazel run //crate_universe/3rdparty:crates_vendor -- --repin`. This should update the `Cargo.lock` file and re-render the BUILD files that allow us to build this code in Bazel.

Reply:
I've tried running that `--repin` command, and it updates the `Cargo.lock` but it's failing with `Error: exit status: 36`.

I'm noticing that the CI builds are failing, too.  It looks like though the `cargo build` binary is getting `cargo_toml` 0.14, but the Bazel built one is still getting 0.13 somehow.  It looks like the vendoring isn't getting updated correctly, but I've not yet been able to figure out how to get it to update cleanly.

Reply:
As a sanity check, I've even tried cloning a fresh `rules_rust` repo without any of my changes and running `bazel run //crate_universe/3rdparty:crates_vendor -- --repin` as the very first command within it.  I still fails with `Error: exit status: 36`.

I'm afraid I'm stuck at this point.

Reply:
I wonder if that's the `cargo_bootstrap_repository` rule failing. I can't know without additional logs but you could try running `cargo update --workspace` in the `crate_universe` directory and trying again.

Reply:
Running `cargo update --workspace` does not produce any change in the files nor in the outcome.  Since this `Error: exit status: 36` failure occurs in a fresh `rules_rust` checkout though, without any of my changes, that's not surprising.

Here are the complete logs of the error I get when I run `bazel run //crate_universe/3rdparty:crates_vendor -- --repin` on a fresh `rules_rust` checkout.  (Once again, this is **without** any of my changes.)

<details>
<summary>Error output</summary>

```
Starting local Bazel server and connecting to it...
WARNING: ignoring _JAVA_OPTIONS in environment.
Loading: 
Loading: 0 packages loaded
Analyzing: target //crate_universe/3rdparty:crates_vendor (1 packages loaded, 0 targets configured)
   Compiling version_check v0.9.4
   Compiling libc v0.2.139
   Compiling proc-macro2 v1.0.49
   Compiling quote v1.0.23
   Compiling unicode-ident v1.0.6
   Compiling syn v1.0.107
   Compiling memchr v2.5.0
   Compiling cfg-if v1.0.0
   Compiling serde_derive v1.0.152
   Compiling serde v1.0.152
   Compiling uncased v0.9.7
   Compiling autocfg v1.1.0
   Compiling thiserror v1.0.38
   Compiling regex-syntax v0.6.28
   Compiling once_cell v1.17.0
   Compiling aho-corasick v0.7.20
   Compiling ppv-lite86 v0.2.17
   Compiling siphasher v0.3.10
   Compiling phf_shared v0.11.1
   Compiling pkg-config v0.3.26
   Compiling log v0.4.17
   Compiling ahash v0.7.6
   Compiling getrandom v0.2.8
   Compiling jobserver v0.1.25
   Compiling regex v1.7.0
   Compiling typenum v1.16.0
Analyzing: target //crate_universe/3rdparty:crates_vendor (97 packages loaded, 198 targets configured)
   Compiling bitflags v1.3.2
   Compiling rand_core v0.6.4
   Compiling cc v1.0.78
   Compiling tinyvec_macros v0.1.0
   Compiling rand_chacha v0.3.1
   Compiling tinyvec v1.6.0
   Compiling libz-sys v1.1.8
   Compiling rand v0.8.5
   Compiling num-traits v0.2.15
   Compiling generic-array v0.14.6
   Compiling percent-encoding v2.2.0
   Compiling thiserror-impl v1.0.38
   Compiling unicode-normalization v0.1.22
   Compiling phf_generator v0.11.1
   Compiling phf_codegen v0.11.1
   Compiling parse-zoneinfo v0.3.0
   Compiling phf v0.11.1
   Compiling num-integer v0.1.45
   Compiling proc-macro-error-attr v1.0.4
   Compiling io-lifetimes v1.0.3
   Compiling serde_json v1.0.91
   Compiling semver v1.0.16
   Compiling unicode-bidi v0.3.8
   Compiling ucd-trie v0.1.5
   Compiling crossbeam-utils v0.8.14
   Compiling idna v0.3.0
   Compiling chrono-tz-build v0.0.3
   Compiling hashbrown v0.12.3
   Compiling form_urlencoded v1.1.0
   Compiling libgit2-sys v0.14.0+1.5.0
   Compiling bstr v0.2.17
   Compiling proc-macro-error v1.0.4
   Compiling unic-char-range v0.9.0
   Compiling fnv v1.0.7
   Compiling unic-common v0.9.0
   Compiling same-file v1.0.6
   Compiling ryu v1.0.12
   Compiling rustix v0.36.6
   Compiling itoa v1.0.5
   Compiling walkdir v2.3.2
   Compiling globset v0.4.9
   Compiling unic-ucd-version v0.9.0
   Compiling unic-char-property v0.9.0
   Compiling url v2.3.1
   Compiling unicode-linebreak v0.1.4
   Compiling chrono-tz v0.6.3
   Compiling thread_local v1.1.4
   Compiling smartstring v1.0.1
   Compiling lazy_static v1.4.0
   Compiling linux-raw-sys v0.1.4
   Compiling iana-time-zone v0.1.53
   Compiling camino v1.1.1
   Compiling chrono v0.4.23
   Compiling ignore v0.4.18
   Compiling unic-ucd-segment v0.9.0
   Compiling crypto-common v0.1.6
   Compiling pest v2.5.2
   Compiling block-buffer v0.10.3
   Compiling anyhow v1.0.68
   Compiling os_str_bytes v6.4.1
   Compiling static_assertions v1.1.0
   Compiling deunicode v0.4.3
   Compiling heck v0.4.0
   Compiling clap_lex v0.3.0
   Compiling clap_derive v4.0.21
   Compiling slug v0.1.4
   Compiling digest v0.10.6
   Compiling pest_meta v2.5.2
   Compiling unic-segment v0.9.0
   Compiling pest_generator v2.5.2
   Compiling git2 v0.15.0
   Compiling globwalk v0.8.1
   Compiling is-terminal v0.4.2
   Compiling num_cpus v1.15.0
   Compiling pest_derive v2.5.2
   Compiling strsim v0.10.0
   Compiling remove_dir_all v0.5.3
   Compiling fastrand v1.8.0
   Compiling smallvec v1.10.0
   Compiling unicode-width v0.1.10
   Compiling humansize v1.1.1
   Compiling home v0.5.4
   Compiling cpufeatures v0.2.5
   Compiling termcolor v1.1.3
   Compiling rustc-hash v1.1.0
   Compiling smawk v0.3.1
   Compiling sha2 v0.10.6
   Compiling cfg-expr v0.12.0
   Compiling tempfile v3.3.0
   Compiling normpath v0.3.2
   Compiling pathdiff v0.2.1
   Compiling indoc v1.0.8
   Compiling textwrap v0.16.0
   Compiling clap v4.0.32
   Compiling toml v0.5.10
   Compiling hex v0.4.3
   Compiling cargo-platform v0.1.2
   Compiling serde_starlark v0.1.8
   Compiling cargo-lock v8.0.3
   Compiling cargo_toml v0.13.3
   Compiling tera v1.17.1
   Compiling cargo_metadata v0.15.2
   Compiling crates-index v0.18.11
   Compiling cargo-bazel v0.6.0 (/home/redacted/tmp/rules_rust2/crate_universe)
    Finished release [optimized] target(s) in 10.63s
Analyzing: target //crate_universe/3rdparty:crates_vendor (97 packages loaded, 198 targets configured)
INFO: Analyzed target //crate_universe/3rdparty:crates_vendor (103 packages loaded, 299 targets configured).
INFO: Found 1 target...
[1 / 6] [Prepa] BazelWorkspaceStatusAction stable-status.txt
[44 / 50] checking cached actions
Target //crate_universe/3rdparty:crates_vendor up-to-date:
  bazel-bin/crate_universe/3rdparty/crates_vendor.sh
INFO: Elapsed time: 16.876s, Critical Path: 0.61s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/crate_universe/3rdparty/crates_vendor.sh --repin
Error: exit status: 36
```
</details>

And this is still on Bazel 6.0 and a `git clone` of `rules_rust` @ commit `85564208e` (release 0.16.1).

Reply:
Odd, I was able to successfully repin on the same commit. Additionally, I was update the pin you were trying to update: https://github.com/bazelbuild/rules_rust/pull/1775

The steps are generally,
1. Update the pin in `//crate_universe:Cargo.toml`
2. From the root of the `rules_rust` checkout, run `cd crate_universe ; cargo update --workspace`
3. Run `bazel run //crate_universe/3rdparty:crates_vendor -- --repin`
4. Update example lockfiles by running `cd examples/crate_universe ; REPIN=1 bazel build ...` (again from the root of the `rules_rust` checkout).

There's another set of lockfiles that need to be updated but I'm planning on removing that: https://github.com/bazelbuild/rules_rust/issues/1776

Reply:
Can you try running the built binary directly? Is Bazel the thing exiting or the script?

Reply:
It's the binary that's exiting.  I extracted the command that `bazel run` and ran I directly and got the same `Error: eixt status: 36` result.

```sh
$ env - BAZEL_REAL=/usr/bin/bazel BUILD_WORKSPACE_DIRECTORY=/home/redacted/tmp/rules_rust2 CARGO=/home/redacted/.cache/bazel/_bazel_redacted/edc61af9f40b4c5aaf23ce66aae36ebf/execroot/rules_rust/bazel-out/k8-fastbuild/bin/crate_universe/3rdparty/crates_vendor.sh.runfiles/rules_rust/../rust_linux_x86_64__x86_64-unknown-linux-gnu__stable_tools/rust_toolchain/bin/cargo RUSTC=/home/redacted/.cache/bazel/_bazel_redacted/edc61af9f40b4c5aaf23ce66aae36ebf/execroot/rules_rust/bazel-out/k8-fastbuild/bin/crate_universe/3rdparty/crates_vendor.sh.runfiles/rules_rust/../rust_linux_x86_64__x86_64-unknown-linux-gnu__stable_tools/rust_toolchain/bin/rustc /home/redacted/.cache/bazel/_bazel_redacted/edc61af9f40b4c5aaf23ce66aae36ebf/execroot/rules_rust/bazel-out/k8-fastbuild/bin/crate_universe/3rdparty/crates_vendor.sh.runfiles/rules_rust/../cargo_bazel_bootstrap/release/cargo-bazel vendor --config /home/redacted/.cache/bazel/_bazel_redacted/edc61af9f40b4c5aaf23ce66aae36ebf/execroot/rules_rust/bazel-out/k8-fastbuild/bin/crate_universe/3rdparty/crates_vendor.sh.runfiles/rules_rust/crate_universe/3rdparty/crates_vendor.cargo-bazel-config.json --splicing-manifest /home/redacted/.cache/bazel/_bazel_redacted/edc61af9f40b4c5aaf23ce66aae36ebf/execroot/rules_rust/bazel-out/k8-fastbuild/bin/crate_universe/3rdparty/crates_vendor.sh.runfiles/rules_rust/crate_universe/3rdparty/crates_vendor.cargo-bazel-splicing-manifest.json --cargo-lockfile /home/redacted/.cache/bazel/_bazel_redacted/edc61af9f40b4c5aaf23ce66aae36ebf/execroot/rules_rust/bazel-out/k8-fastbuild/bin/crate_universe/3rdparty/crates_vendor.sh.runfiles/rules_rust/crate_universe/Cargo.lock --buildifier /home/redacted/.cache/bazel/_bazel_redacted/edc61af9f40b4c5aaf23ce66aae36ebf/execroot/rules_rust/bazel-out/k8-fastbuild/bin/crate_universe/3rdparty/crates_vendor.sh.runfiles/rules_rust/../cargo_bazel.buildifier-linux-amd64/file/buildifier.exe --repin
Error: exit status: 36
```

Adding `RUST_LOG=trace` and `RUST_BACKTRACE=1` to the environment did not produce any additional output, sadly.

Reply:
What happens if you run just the built `cargo_bazel` binary? 

```
/home/redacted/.cache/bazel/_bazel_redacted/edc61af9f40b4c5aaf23ce66aae36ebf/execroot/rules_rust/bazel-out/k8-fastbuild/bin/crate_universe/3rdparty/crates_vendor.sh.runfiles/rules_rust/../cargo_bazel_bootstrap/release/cargo-bazel --help
```

Reply:
It will print the usage information.  So we can run the binary directly, but running the exact command `bazel run` invokes seems to produce the error.

<details>
<summary>Command output</summary>

```sh
$ /home/redacted/.cache/bazel/_bazel_redacted/edc61af9f40b4c5aaf23ce66aae36ebf/execroot/rules_rust/bazel-out/k8-fastbuild/bin/crate_universe/3rdparty/crates_vendor.sh.runfiles/rules_rust/../cargo_bazel_bootstrap/release/cargo-bazel --help
crate_universe` is a collection of tools which use Cargo to generate build targets for Bazel.

Usage: cargo-bazel <COMMAND>

Commands:
  generate  Generate Bazel Build files from a Cargo manifest
  splice    Splice together disjoint Cargo and Bazel info into a single Cargo workspace manifest
  query     Query workspace info to determine whether or not a repin is needed
  vendor    Vendor BUILD files to the workspace with either repository definitions or `cargo vendor` generated sources
  help      Print this message or the help of the given subcommand(s)

Options:
  -h, --help     Print help information
  -V, --version  Print version information
```
</details>

Reply:
Can you try slowly adding back flags to see if any more revealing info can be uncovered?

Reply:
I found what appears to be causing it: the `env -` prefix.  If I run the command without that `env -` prefix, it will regenerate the vendor files properly.

It seems like the program is relying on some environment variable other than the ones explicitly provided.  Though I have no clue which one.

Reply:
For kicks, I tried changing how the `cargo_vendor` script is generated with the following patch, and it resolves the `Error: exit status: 36` issue.

```diff
diff --git a/crate_universe/private/crates_vendor.bzl b/crate_universe/private/crates_vendor.bzl
index 4d45c009..68a80a77 100644
--- a/crate_universe/private/crates_vendor.bzl
+++ b/crate_universe/private/crates_vendor.bzl
@@ -12,7 +12,7 @@ export RUNTIME_PWD="$(pwd)"
 if [[ -z "${{BAZEL_REAL:-}}" ]]; then
     BAZEL_REAL="$(which bazel || echo 'bazel')"
 fi
-eval exec env - BAZEL_REAL="${{BAZEL_REAL}}" BUILD_WORKSPACE_DIRECTORY="${{BUILD_WORKSPACE_DIRECTORY}}" {env} \\
+eval exec env BAZEL_REAL="${{BAZEL_REAL}}" BUILD_WORKSPACE_DIRECTORY="${{BUILD_WORKSPACE_DIRECTORY}}" {env} \\
 "{bin}" {args} "$@"
 """
 ```

Reply:
Seems like there's additional environment variables on your host that are needed? Do you know what they are?

Reply:
No idea at all. :smiling_face_with_tear:

Reply:
Reframing the question, is there a strong reason why this script needs to clear out all other environment variables before running the binary?  That seems likely to break things one way or another down the line, even if we could find the specific environment variable that it needs from my system.

Reply:
The original idea was that the results of running this target should be the same, regardless of what system it's run on. If someone generates dependencies and ends up with some output unique to their environment, it could lead to confusion when the next person with a different environment updates dependencies. Assuming you have a frozen crates registry, it should always do the same thing, no matter what. Thus, we eliminate environment variables we don't know we need.

Reply:
In principle, that makes sense.  But in practice the `cargo-bazel` binary has a large and deep dependency tree, and if any of your dependencies expect, rely on, or react to an environment variable you aren't aware of, then the tooling breaks, as it has for me.

To complicate matters further, the rest of the processes being run by Bazel are running with a full environment; so by calling `cargo-bazel` for vendoring with `env -`, it's become a special case that runs with an environment different from everything else Bazel has been running.

I hear where you're coming from, but since all those other Bazel processes are exposed to the user's full environment anyway, you can't really shield yourself from its effects on the build as a whole.

If you have a suggestion for how I could isolate the particular environment variable(s) that `cargo-bazel vendor` needs from my system, I'm game to try, but I'd still caution that you'll probably wind up back in this situation again down the line with someone else one day.

Reply:
Found the environment variable that `cargo-bazel` needed: it was `PATH`.  And it didn't even need anything strange or special from my `PATH`; just adding `PATH=/usr/bin` to the `env` invocation was enough to get rid of the `exit status: 36` error.

So if you really want to keep a limited environment, you could consider adding `"PATH=${PATH}"` to your `env` invocation.

Reply:
Do you know what in `PATH` was required? I wouldn't have expected `cargo-bazel` to be invoking anything in `PATH`.

Reply:
It doesn't look like `cargo-bazel` itself is relying on `PATH`.  But `cargo-bazel` does call out to `/usr/bin/bazel info output_base`, and it's the `bazel` command itself that's failing and causing the `36` exit code.

I can reproduce that smaller kernel of the problem as follows:

```sh
$ env - /usr/bin/bazel info output_base
FATAL: Could not find system javabase. Ensure JAVA_HOME is set, or javac is on your PATH.

$ env - PATH=/usr/bin /usr/bin/bazel info output_base
Starting local Bazel server and connecting to it...
/home/redacted/.cache/bazel/_bazel_redacted/18ef78106ff50771ab6492a671b0fdf0
```